airbyte-integrations/connectors/source-amplitude/source_amplitude/api.py

```diff
@@ -0,0 +1,205 @@
+        yield from respose_data.get(self.name, [])
+
+    def path(self, **kwargs) -> str:
+        return f"/{self.api_version}/{self.name}"
+
+
+class Cohorts(AmplitudeStream):
+    primary_key = "id"
+    api_version = 3
+
+
+class Annotations(AmplitudeStream):
+    primary_key = "id"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        respose_data = response.json()
+        yield from respose_data.get("data", [])
+
+
+class IncrementalAmplitudeStream(AmplitudeStream, ABC):
+    state_checkpoint_interval = 100
+    base_params = {}
+    cursor_field = "date"
+    date_template = "%Y%m%d"
+
+    def __init__(self, start_date: str, **kwargs):
+        super().__init__(**kwargs)
+        self._start_date = pendulum.parse(start_date)
+
+    @property
+    @abstractmethod
+    def time_interval(self) -> dict:
+        """
+        Defining the time interval to determine the difference between the end date and the start date.
+        """
+        pass
+
+    def _get_end_date(self, current_date: pendulum, end_date: pendulum = pendulum.now()):
+        if current_date.add(**self.time_interval).date() < end_date.date():
+            end_date = current_date.add(**self.time_interval)
+        return end_date
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        latest_benchmark = latest_record[self.cursor_field]
+        if current_stream_state.get(self.cursor_field):
+            return {self.cursor_field: max(latest_benchmark, current_stream_state[self.cursor_field])}
+        return {self.cursor_field: latest_benchmark}
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        parsed = urlparse.urlparse(response.url)
+        end = parse_qs(parsed.query).get("end", None)
+        if end:
+            end_time = pendulum.parse(end[0])
+            now = pendulum.now()
+            if end_time.date() < now.date():
+                return {"start": end[0], "end": self._get_end_date(end_time).strftime(self.date_template)}
+        return None
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = self.base_params
+        if next_page_token:
+            params.update(next_page_token)
+        else:
+            start_datetime = self._start_date
+            if stream_state.get(self.cursor_field):
+                start_datetime = pendulum.parse(stream_state[self.cursor_field])
+
+            params.update(
+                {
+                    "start": start_datetime.strftime(self.date_template),
+                    "end": self._get_end_date(start_datetime).strftime(self.date_template),
+                }
+            )
+        return params
+
+
+class Events(IncrementalAmplitudeStream):
```

Comment:
are events returned in ascending order of the cursor field? if so, why don't we checkpoint using checkpoint_interval? This would allow the connector to resume where it left off even if it failed halfway through the sync.

Reply:
same question for other streams BTW

Reply:
Also, this would happen automatically if we override the `stream_slices` method. Because you are doing time slicing here, it's a good fit for slicing and keeps track of state automatically.

Reply:
I used `state_checkpoint_interval`, it was equal to 100 for all Incremental streams, but now I slightly corrected, for all I set 10, and for Events I set 1000 (since there are quite a few entries)